diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/chapter3/files/NewlinesRule.kt

```diff
@@ -491,35 +495,116 @@ class NewlinesRule(configRules: List<RulesConfig>) : DiktatRule(
-                "$entryType should be placed on different lines in declaration of <${node.getParentIdentifier()}>"
-            } ?: "$entryType should be placed on different lines"
-            WRONG_NEWLINES.warnAndFix(configRules, emitWarn, isFixMode,
-                warnText, node.startOffset, node) {
-                invalidCommas.forEach { commaOrRpar ->
-                    val nextWhiteSpace = commaOrRpar.treeNext?.takeIf { it.elementType == WHITE_SPACE }
-                    if (commaOrRpar.elementType == COMMA) {
-                        nextWhiteSpace?.treeNext?.let {
-                            commaOrRpar.appendNewlineMergingWhiteSpace(nextWhiteSpace, nextWhiteSpace.treeNext)
-                        } ?: commaOrRpar.treeNext?.treeParent?.appendNewlineMergingWhiteSpace(nextWhiteSpace, commaOrRpar.treeNext)
-                    } else {
-                        commaOrRpar.treeParent?.appendNewlineMergingWhiteSpace(nextWhiteSpace, commaOrRpar)
+
+        val classDefinitionNode = valueParameterList[0].treeParent?.treeParent
+        val colonNodes = classDefinitionNode?.getAllChildrenWithType(COLON)
+        val newlineBeforeSuperTypeList = colonNodes?.find {colonNode ->
+            val newlineNode = colonNode.treeNext
+            val superClassType = newlineNode?.treeNext
+            newlineNode?.text?.count { it == '\n' } == 1 && superClassType?.elementType == SUPER_TYPE_LIST
+        }
+
+        // list elements are correct if they are sequence of (COMMA, '\n', supertype)
+        var areElementsCorrect = true
+        var valueParameter = valueParameterList[0].treeNext
+        while (valueParameter != null) {
+            val newlineNode = valueParameter.treeNext
+            val superClassType = valueParameter.treeNext?.treeNext
+            if (valueParameter.elementType != COMMA || newlineNode?.text?.count { it == '\n' } != 1 ||
+                    superClassType?.elementType != SUPER_TYPE_ENTRY) {
+                areElementsCorrect = false
+                break
+            }
+            valueParameter = superClassType?.treeNext
+        }
+        return newlineBeforeSuperTypeList != null && areElementsCorrect
+    }
+
+    private fun setSuperClassesOnSeparateLines(
+        node: ASTNode,
+        valueParameterList: List<ASTNode>,
+        warnText: String,
+        colonNodes: List<ASTNode>?
+    ) {
+        WRONG_NEWLINES.warnAndFix(
+            configRules, emitWarn, isFixMode,
+            warnText, node.startOffset, node
+        ) {
+            valueParameterList.forEach { superClassNode ->
+                val commaNode = superClassNode.treeNext
+                val whiteSpaceNode = commaNode?.treeNext
+                // put super classes on separate lines
+                if (superClassNode.elementType in listOf(SUPER_TYPE_CALL_ENTRY, SUPER_TYPE_ENTRY) &&
+                        commaNode != null && whiteSpaceNode?.elementType == WHITE_SPACE
```

Comment:
should we also check, that `comaNode.text == ","`?

Reply:
Done